### ISSUE TITLE
false positive: unsupported-membership-type

### ISSUE THREAD

**Issue by trahman1318** [PyCQA/pylint#8313]
### Bug description

```python
import subprocess
import unittest


def execute_thing(args):
    return subprocess.run(args, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE).stdout


class TestThing(unittest.TestCase):
    def test_thing(self):
        args = [
            "test",
        ]
        output = execute_thing(args)

        expected = "expected output"
        assert expected in output  # false positive
```

### Command used

```shell
pylint -E myfile.py
```

### Pylint output

```shell
************* Module myfile
myfile.py:19:27: E1135: Value 'output' doesn't support membership test (unsupported-membership-test)
```

### Expected behavior

I expect no linting errors because `str.__contains__()` exists.

### Pylint version

```shell
pylint 2.7.2
astroid 2.5.8
Python 3.10.4 (main, Dec  9 2022, 18:30:25) [GCC 11.3.0]
```

### OS / Environment

Ubuntu 18.04

### Additional dependencies

Package           Version
----------------- -------
astroid           2.5.8
isort             5.12.0
lazy-object-proxy 1.9.0
mccabe            0.6.1
pip               22.0.4
pylint            2.7.2
setuptools        58.1.0
toml              0.10.2
wrapt             1.12.1

**Comment by Pierre-Sassoulas** [comment]
I cannot reproduce. 2.7.2, is really old version of pylint. There's been ~= 3250 commits on main since it released. If I had to guess you're using the debian version (probably from buster or even stretch instead of bullseyes). Could you upgrade to the latest version with pip ?

**Comment by jacobtylerwalls** [comment]
Feel free to reopen if you can reproduce against the development version of pylint.

**Comment by davetapley** [comment]
Not sure if same as OP, but I have a repro for this (with a caveat):

I'm using `extension-pkg-allow-list = ["falcon"]` because https://github.com/falconry/falcon/issues/1553#issuecomment-545284240

Then a pattern is to [subclass](https://falcon.readthedocs.io/en/stable/api/app.html) `falcon.Request`, e.g.:

```py
from falcon.request import Request

class MyRequest(Request):
    def my_function(self):
        if '.parquet' in self.path:
            ...
```

but that causes:
```
E1135: Value 'self.path' doesn't support membership test (unsupported-membership-test)
```

even though `path` is [definitely a string](https://github.com/falconry/falcon/blob/1ef41db2ea17eb0015962c3f22668796de27dcd1/falcon/request.py#L159).

---

I _assume_ root cause is the `extension-pkg-allow-list` preventing `pylint` from 'seeing' the `path: str` inside `falcon.Request`?

Perhaps something similar was happening with OP and `subprocess`?

Happy to open a new issue if that's easier.

**Comment by Pierre-Sassoulas** [comment]
Pylint doesn't use type information it infers from actual value used in the code.

**Comment by Pierre-Sassoulas** [comment]
(It's an intuition, I didn't check in detail but,) it's very often due to code that is "too dynamic" i.e. you have globals based on environment variables, conditionals or even conditionals in loop that astroid's inference cannot follow.
